packages/data/src/data-manager.ts

```diff
@@ -286,111 +289,120 @@ export class DataManager implements DataManagerInterface {
             this._loggerManager?.info?.(
               'DataManager.matchRulesByField()',
-              MESSAGES.AUDIENCE_NOT_RESTRICTED
+              MESSAGES.NON_PERMANENT_AUDIENCE_NOT_RESTRICTED
             );
           }
-        }
-        // Get attached segmentation audiences
-        segments = this.getItemsByIds(
-          experience.audiences,
-          'segments'
-        ) as Array<ConfigSegment>;
-        if (segments.length) {
-          // Validate custom segments against segmentations
-          matchedSegments = this.filterMatchedCustomSegments(
-            segments,
-            visitorId
+        } else {
+          audiencesMatched = true; // Empty experience audiences list means no restrictions
+          this._loggerManager?.info?.(
+            'DataManager.matchRulesByField()',
+            MESSAGES.AUDIENCE_NOT_RESTRICTED
           );
-          if (matchedSegments.length) {
-            for (const item of matchedSegments) {
-              this._loggerManager?.info?.(
-                'DataManager.matchRulesByField()',
-                MESSAGES.SEGMENTATION_MATCH.replace('#', item?.[identityField])
-              );
-            }
-          }
         }
-        // If there are some matched audiences
-        if (
-          !visitorProperties ||
-          !audiencesToCheck.length || // Empty audiences list means there's no restriction for the audience
-          (audiencesToCheck.length && matchedAudiences.length) ||
-          (segments.length && matchedSegments.length)
-        ) {
-          // And experience has variations
-          if (experience?.variations && experience?.variations?.length) {
+      }
+      // Get attached segmentation audiences
+      segments = this.getItemsByIds(
+        experience.audiences,
+        'segments'
+      ) as Array<ConfigSegment>;
+      if (segments.length) {
+        // Validate custom segments against segmentations
+        matchedSegments = this.filterMatchedCustomSegments(segments, visitorId);
+        if (matchedSegments.length) {
+          for (const item of matchedSegments) {
             this._loggerManager?.info?.(
               'DataManager.matchRulesByField()',
-              MESSAGES.EXPERIENCE_RULES_MATCHED
-            );
-            return experience;
-          } else {
-            this._loggerManager?.debug?.(
-              'DataManager.matchRulesByField()',
-              MESSAGES.VARIATIONS_NOT_FOUND,
-              this._mapper({
-                visitorProperties: visitorProperties,
-                audiences: audiences
-              })
+              MESSAGES.SEGMENTATION_MATCH.replace('#', item?.[identityField])
             );
           }
+        }
+        segmentsMatched = Boolean(matchedSegments.length);
+      }
+      // If there are some matched locations and/or audiences
+      if (locationMatched || audiencesMatched || segmentsMatched) {
```

Comment:
I still do not think it's ok @abbaseya for two reasons:
1. for segments, you start with `segmentsMatched = false` and you are not making it true in case experience has no segments 
2. `audiencesMatched || segmentsMatched `would let the experience run if either of the two is true which is not what we want. We need to make sure both are matched . So probably this condition should become `if (audiencesMatched && segmentsMatched) {` - removing locationMatched  since that was already checked previously(above)